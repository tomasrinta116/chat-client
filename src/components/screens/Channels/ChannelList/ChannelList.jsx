import React from "react";
import List from "@material-ui/core/List";
import ListSubheader from "@material-ui/core/ListSubheader";
import { withStyles } from "@material-ui/core/styles";
import Button from "@material-ui/core/Button";
import Paper from "@material-ui/core/Paper";
import styles from "./ChannelList.styles";
import PropTypes from "prop-types";

class ChannelList extends React.PureComponent {
  handleClick = channelURL => async event => {
    const { history, enterChannel } = this.props;
    await enterChannel(channelURL);
    history.push(`/chat/${channelURL}`);
  };

  render() {
    const {
      classes: { channelListContainer, channelButtonList, channelButton },
      channels
    } = this.props;
    return (
      <div className={channelListContainer}>
        <Paper>
          <List
            className={channelButtonList}
            subheader={<ListSubheader color="primary">Channels</ListSubheader>}
          >
            {channels.map((channel, index) => {
              const { name, url } = channel;
              return (
                <div key={name + index.toString()}>
                  <Button
                    data-testid={`${name}${index.toString()}`}
                    onClick={() => this.handleClick(url)()}
                    className={channelButton}
                  >
                    {name}
                  </Button>
                </div>
              );
            })}
          </List>
        </Paper>
      </div>
    );
  }
}

ChannelList.propTypes = {
  channels: PropTypes.arrayOf(PropTypes.object).isRequired,
  history: PropTypes.object.isRequired,
  enterChannel: PropTypes.func.isRequired
};

export default withStyles(styles)(ChannelList);
