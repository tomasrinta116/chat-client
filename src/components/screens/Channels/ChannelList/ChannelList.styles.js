const styles = {
  channelListContainer: {
    textAlign: "center"
  },
  channelButtonList: {
    maxHeight: "70vh",
    overflowY: "auto"
  },
  channelButton: {
    width: "100%",
    padding: "1rem"
  }
};

export default styles;
