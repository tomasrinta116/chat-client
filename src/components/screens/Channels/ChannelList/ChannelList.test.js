import React from "react";
import { shallow } from "enzyme";
import sinon from "sinon";
import ChannelList from "./ChannelList";
import List from "@material-ui/core/List";
import Button from "@material-ui/core/Button";

describe("<ChannelList />", () => {
  const props = {
    channels: [
      { name: "test-channel1", url: "www.test1.com" },
      { name: "test-channel2", url: "www.test2.com" },
      { name: "test-channel3", url: "www.test3.com" }
    ], 
    history: {},
    enterChannel: jest.fn(),
  };

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ChannelList {...props} />).dive();
  });

  test("it renders correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("it renders the correct number of items", () => {
    const list = wrapper.find(List);
    expect(list).toHaveLength(1);
    const buttons = wrapper.find(Button);
    expect(buttons).toHaveLength(3);
  });

  test("handleClick is called when clicking on a list item", () => {
    const spy = sinon.spy(wrapper.instance(), "handleClick");
    const button1 = wrapper.find({ "data-testid": "test-channel10" }).dive();
    expect(button1).toHaveLength(1);
    console.log(button1.debug());
    button1.simulate("click");
    expect(spy.called).toBe(true);
  });

});
