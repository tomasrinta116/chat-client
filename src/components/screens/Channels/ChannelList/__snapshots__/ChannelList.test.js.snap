// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChannelList /> it renders correctly 1`] = `
<div
  className="ChannelList-channelListContainer-1"
>
  <WithStyles(Paper)>
    <WithStyles(List)
      className="ChannelList-channelButtonList-2"
      subheader={
        <WithStyles(ListSubheader)
          color="primary"
        >
          Channels
        </WithStyles(ListSubheader)>
      }
    >
      <div
        key="test-channel10"
      >
        <WithStyles(Button)
          className="ChannelList-channelButton-3"
          data-testid="test-channel10"
          onClick={[Function]}
        >
          test-channel1
        </WithStyles(Button)>
      </div>
      <div
        key="test-channel21"
      >
        <WithStyles(Button)
          className="ChannelList-channelButton-3"
          data-testid="test-channel21"
          onClick={[Function]}
        >
          test-channel2
        </WithStyles(Button)>
      </div>
      <div
        key="test-channel32"
      >
        <WithStyles(Button)
          className="ChannelList-channelButton-3"
          data-testid="test-channel32"
          onClick={[Function]}
        >
          test-channel3
        </WithStyles(Button)>
      </div>
    </WithStyles(List)>
  </WithStyles(Paper)>
</div>
`;
